FEATURE: Consider PHP attributes in proxy method building

robertlemke opened the pull request:
Added support for preserving PHP 8 attributes in generated proxy class methods. This feature enables correct argument passing from attributes to proxied methods which allows developers to use attributes instead of annotations in most cases.

Resolves #3075

robertlemke committed:
FEATURE: Consider PHP attributes in proxy method building

Added support for preserving PHP 8 attributes in generated proxy class methods. This feature enables correct argument passing from attributes to proxied methods which allows developers to use attributes instead of annotations in most cases.

Resolves #3075

kitsunet commented on Neos.Flow/Classes/ObjectManagement/Proxy/ProxyMethodGenerator.php:
```diff
@@ -174,4 +242,78 @@ protected function buildCallParentMethodCode(string $fullClassName, string $meth
+        }
+
+        // Fallback for any other types (shouldn't happen with PHP attributes)
+        return '';
```
Do we want to except here instead? Just to be sure so that potential future expansions of the syntax get to an exception instead of resulting in weird proxies?

I am probably fine either way, but I could imagine e.g. floats being added which would then fall through to here...

robertlemke replied:
Yes, good point, I pushed a commit handling that and added support for floats.

kitsunet commented on Neos.Flow/Classes/ObjectManagement/Proxy/ProxyMethodGenerator.php:
```diff
@@ -174,4 +242,78 @@ protected function buildCallParentMethodCode(string $fullClassName, string $meth
+        $formattedArguments = [];
+
+        foreach ($arguments as $key => $value) {
+            $formattedArguments[] = "{$key}: " . $this->formatAttributeValue($value);
```
According to the docs something like this here is valid:

`#[JsonSerialize('foobar')]`

so we probably need to have a test for this kind of value without key and properly handle it here.

robertlemke replied:
Thanks! I added support for that, too.

kitsunet commented on Neos.Flow/Tests/Functional/ObjectManagement/Fixtures/ClassWithPhpAttributes.php:
```diff
@@ -21,4 +21,12 @@
+    /**
+     * @return void
+     */
+    #[Flow\Around(pointcutExpression: "method(somethingImpossible())")]
```
Now I wonder, I think we always create the "Flow" use statement in proxy classes, but what about other namespaces, do you know if we copy all use statements over to the proxy, otherwise this might fail?

robertlemke replied:
This is independent from any imports, the proxy method uses a fully qualified attribute name including the absolute namespace.

kitsunet replied:
oh I see, neat then!

robertlemke committed:
TASK: Throw exception on unsupported attribute value type

If a class method being proxied contains an attribute value of an unsupported type, an exception is thrown at compile time.

robertlemke committed:
TASK: Add support for positional attribute arguments

Proxy methods using attributes with positional arguments are now rendered correctly.

kitsunet reviewed (approved):
Very nice, thanks for the quick additions :)

kdambekalns commented on Neos.Flow/Classes/ObjectManagement/Proxy/ProxyMethodGenerator.php:
```diff
@@ -83,7 +100,59 @@ public function generate(): string
+            . $this->getName() . '(';
+
+        $output .= implode(', ', array_map(
+            static fn(ParameterGenerator $parameter): string => $parameter->generate(),
```
```suggestion
            static fn (ParameterGenerator $parameter): string => $parameter->generate(),
```

kdambekalns commented on Neos.Flow/Classes/ObjectManagement/Proxy/ProxyMethodGenerator.php:
```diff
@@ -174,4 +243,93 @@ protected function buildCallParentMethodCode(string $fullClassName, string $meth
+                $this->getFullOriginalClassName(),
+                $methodName,
+                get_debug_type($value)
+            ), 1705501433
```
```suggestion
            ),
            1705501433
```

kdambekalns commented on Neos.Flow/Tests/Functional/ObjectManagement/ProxyCompilerTest.php:
```diff
@@ -23,6 +25,8 @@
 use Neos\Flow\Tests\Functional\ObjectManagement\Fixtures\PHP81\BackedEnumWithMethod;
 use Neos\Flow\Tests\Functional\ObjectManagement\Fixtures\PrototypeClassA;
 use Neos\Flow\Tests\Functional\ObjectManagement\Fixtures\PrototypeClassK;
+use Neos\Flow\Tests\Functional\ObjectManagement\Fixtures\SampleAttribute;
```
```suggestion
```

kdambekalns reviewed (commented):
StyleCI fixes…

kdambekalns committed:
TASK: Fix StyleCI complaints

kdambekalns reviewed (approved):
👍 by 👀
